### ISSUE TITLE
Reduce allocations in (De-)SerializeCompact

### ISSUE THREAD

**Issue by gterzian** [massalabs/massa#2264]
I think one of the most obvious potential optimization of (de-)serialization would be to avoid unnecessarily allocating on the heap(when using `Vec<u8>`).

For example, inside [`to_bytes_compact`](https://github.com/massalabs/massa/blob/b1240784b2984cad46a1e0de72b915b2448504ce/massa-models/src/block.rs#L257) for `Block`, `to_bytes_compact` is called on each `Operation` in the block. 

https://github.com/massalabs/massa/blob/b1240784b2984cad46a1e0de72b915b2448504ce/massa-models/src/block.rs#L273

Each call to `operation.to_bytes_compact()` allocates a vector, and then immediately drops it. 

I would say we should avoid these recursive calls to `to_bytes_compact`, and instead work with array/slices when possible to avoid heap allocations. 

For example, when we serialize a [`Message`](https://github.com/massalabs/massa/blob/b1240784b2984cad46a1e0de72b915b2448504ce/massa-network/src/messages.rs#L70) to send it over the network, we actually just need a `&[u8]`. 

Currently when sending a block over the network, we pass the `Message` around and call [`msg.to_bytes_compact()?`](https://github.com/massalabs/massa/blob/b1240784b2984cad46a1e0de72b915b2448504ce/massa-network/src/binders.rs#L41), which allocates vectors for:

1. The message,
2. The block,
3. each operation in the block
4. and so on...

And all of this just to do [`write_half.write_all(&bytes_vec[..])`](https://github.com/massalabs/massa/blob/b1240784b2984cad46a1e0de72b915b2448504ce/massa-network/src/binders.rs#L54).

----

A solution would be to "unwrap" those recursive serialization calls, and instead do everything at the point of use in network, using arrays.

**Comment by gterzian** [comment]
Once place were we would need a vector, would be if we were to cache the serialized representation, as current done in https://github.com/massalabs/massa/pull/2233.

And even then we only need one vector for the "top-level" structure to be cached, not one for each internal call to `to_bytes_compact`.

**Comment by damip** [comment]
Here is a proposal based on some ideas I had, as well as your comments:

# Serialization refactoring

## General principles

* avoid copies and allocations (especially on the heap)
* have rich deserialized objects that also contain their serialized versions of that object that are useful for sending through the network, so that we don't have to re-serialize 
* use the nom library for safety and ease of serialization/parsing instead of doing it manually: https://docs.rs/nom/latest/nom/

## Block

```rust


In storage:

struct Endorsement {
    version: EndorsementVersion,
    creator_pubkey: PublicKey,
    creator_address: Address,
    target_slot: Slot,
    target_block: BlockId
}

struct BlockPayload {
    endorsements
    denunciations
    operations
}

struct Block {
    block_id: BlockId,
    creator_pubkey: PublicKey,
    creator_address: Address,
    slot: Slot,
    parents: Vec<BlockId>,
    payload_hash: Hash
    
    serialized_header: [u8], // serialized version of the block header
    serialized_payload: Option<[u8]>,  // serialized version of BlockPayload
    
    payload: Option<BlockPayload>
}


```


### serialized_header format

Note that the public key is eluded and shoud be retrieved from PoS for that slot.  (pos draws should be pubkeys)


```
block ID:
    0 - version
    1 - hash(rest of the serialized data)
2 - slot
3 - parent block IDs (concatenated without length marker)
4 - payload_hash (hash of endorsements and operations)
5 - signature(hash([version + slot + hash(parents + payload_hash)]))

```

How we deserialize when receiving:

* start by deserializing the block ID (version + hash(contentHash + signature))
  * fail if format invalid or version unsupported
* if we already know about this block ID
    * mark the sender as knowing about the block (ban if the block is invalid)
    * stop processing the block
* compute the block ID by hashing the rest of the serialized data, ensure it matches the advertised block ID, otherwise ban sender
* mark the sender as knowing the block
* deserialize data. If failure => ignore, ban sender
* get PoS draw for slot (directly a public key). If fail (too much in the future/past) => ignore block
* verify signature (also requires reserializing some stuff and computing hashes). If fail => ignore, ban sender


### serialized_payload format

TODO

**Comment by AureliaDolo** [comment]
in block payload
* the structs are vecs ?
* the endorsements are no longer in the header?

in block 
* why the payload is optional ? it can always be just empty vecs ?
* when deserializing the endorsements, de we retrieve the public key from pos ?

Can you make a diagram of that ?

**Comment by damip** [comment]
@AureliaDolo

> the structs are vecs ?

 updated the spec

> the endorsements are no longer in the header?

because we might not need them for the checks we do before asking for the full block

> why the payload is optional ? it can always be just empty vecs ?

because we might not have received it yet. There needs to be a difference between "we don't have it yet" and "we have it but it has zero operations". Updated the spec

> when deserializing the endorsements, de we retrieve the public key from pos ?

updated the spec

**Comment by gterzian** [comment]
> How we deserialize when receiving:

Would we store the block in storage when this series of steps has been successful? The other point would be when creating a block locally. Correct?

```
struct BlockPayload {
    serialized: [u8]  // serialized version of the payload
    endorsements: Vec<Option<Endorsement>>,
    denunciations: Vec<Denunciation>
    operations: Vec<Operation>
}
```

So the `BlockHeaderContent` would not include the current ` Vec<Endorsement>`?

**Comment by gterzian** [comment]
Also, do we really need to store the serialized forms into the objects themselves? I would be in favor of keeping it separate, see the currently proposed [`StoredBlock`](https://github.com/massalabs/massa/pull/2233/files#diff-dc6d1c306121e17c2c4f8de4f32d293dc7bd8776b5da3372f172e0a98962a280R6). 

Mainly because if the serialized representation is stored on the object itself, I think this would require using an option, and also I wonder if it would make sense to include it in the network message.

In other words, the stored serialized form seems best treated as separate from the object itself, and store "alongside" it in storage.

**Comment by gterzian** [comment]
> Do we really need to store the serialized forms into the objects themselves? I would be in favor of keeping it separate, see the currently proposed [StoredBlock](https://github.com/massalabs/massa/pull/2233/files#diff-dc6d1c306121e17c2c4f8de4f32d293dc7bd8776b5da3372f172e0a98962a280R6).

By the way, I think we should not so much store the serialized objects, but rather the serialized network messages. I actually initially made a mistake in the PR I linked to, and corrected it [here](https://github.com/massalabs/massa/pull/2233/commits/917258e58519265421a8dd2fc70a17a3795e8013), and as I think can be seen, the most straightforward thing to do is to store and re-use the message itself, not the object. 

There may be some exceptions, for example when we export objects, however these might not be "in the critical path" in the same way as the regular network messages. 

Anyway, to be considered, and from my experience over at https://github.com/massalabs/massa/pull/2233 I am getting more convinced that the caching of serialized objects should be kept as separate from the object itself as possible, as opposed to what has been sketched above where the cache is inside the object itself.

**Comment by damip** [comment]
> Do we really need to store the serialized forms into the objects themselves? I would be in favor of keeping it separate, see the currently proposed [`StoredBlock`](https://github.com/massalabs/massa/pull/2233/files#diff-dc6d1c306121e17c2c4f8de4f32d293dc7bd8776b5da3372f172e0a98962a280R6).

Both work. I went for a "flat" version simply allows using a lower number of nested sub-structures. 

> Mainly because if the serialized representation is stored on the object itself, I think this would require using an option, and also I wonder if it would make sense to include it in the network message, so you'd have to set it to `None` first and do other kinds of extra bookkeeping.

This is how I envisioned it:

When receiving a block: see proposed optimized algorithm above

When creating a block:
* call a constructor function that takes the block properties as a structure (everything except BlockID, signature, serialized etc...). This function assembles the block, serializes the block contents, computes the signature, computes and populates the serialized version, computes and populates the block ID etc... 

> In other words, the stored serialized form seems best treated as separate from the object itself, and stored "alongside" it in storage, as opposed to it being part of the object itself.

Maybe. I just didn't catch a good argument in favor of the added complexity of carrying the serialized version "on the side" (more structures that will never be written outside of block reception/creation, and/or more function arguments, more "dots" resulting in stuff like stored_block.block.header.contents etc. everywhere)

**Comment by damip** [comment]
> `Option<BlockPayload>` I'm also unsure about this concept. It seems to be sort of the equivalent of the "block" currently, where in the case of it being`None` the `Block` would be the equivalent of a current "header".
> I find the concept of headers and full blocks more intuitive than that of a block with an optional "payload". At best both structure are equivalently clear and so it's unclear why there is a need to refactor it that way.

That's mostly to prepare for the extra flexibility needed by the multistake protection. Some (multistaked) blocks could be considered complete even without a payload (which we could explicitly discard a posteriori or avoid downloading in some cases). 

> The one thing that could make sense in my opinion is moving the endorsements out of the current headers, to the block, since that is where the other data is found as well.

Yes that's pretty nice and also fits well with the concept of "payload" that could be further extended in the future if more "non-graph" data is added to blocks. However I think we might need to keep denunciations outside of the payload (still thinking about it)

**Comment by damip** [comment]
> > Do we really need to store the serialized forms into the objects themselves? I would be in favor of keeping it separate, see the currently proposed [StoredBlock](https://github.com/massalabs/massa/pull/2233/files#diff-dc6d1c306121e17c2c4f8de4f32d293dc7bd8776b5da3372f172e0a98962a280R6).

See my response above

> By the way, I think we should not so much store the serialized objects, but rather the serialized network messages. I actually initially made a mistake in the PR I linked to, and corrected it [here](https://github.com/massalabs/massa/pull/2233/commits/917258e58519265421a8dd2fc70a17a3795e8013), and as I think can be seen, the most straightforward thing to do is to store and re-use the message itself, not the object.

I don't like the idea of storing network messages. What if we decide to batch header transmission like we do for operations, and send different batches to different people ? What about future versioning, with different nodes with different block and network protocol versions coexisting ?

Also, we could avoid re-sending the block header by making a `BlockPayload` message to send only the missing payload.

Also, maybe this could be a nicer way of accessing the block (to avoid copying the whole block when we just want one field):

```
storage
  .get_block(&block_id) // returns an `Option<RwLockReadGuard<Block>>`
  .expect("blablabla")
  .serialized
  .clone()
```

> There may be some exceptions, for example when we export objects, however these might not be "in the critical path" in the same way as the regular network messages.
> Anyway, to be considered, and from my experience over at #2233 I am getting more convinced that the caching of serialized objects should be kept as separate from the object itself as possible, as opposed to what has been sketched above where the cache is inside the object itself.

We should discuss this more in depth I think because I didn't catch the fundamental reason.

**Comment by gterzian** [comment]
> Both work. I went for a "flat" version simply allows using a lower number of nested sub-structures.

I don't understand this, because the current proposal appears "nested" to me, since the serialized form appears to be nested inside the block structure. "Flat" would be having separate concepts alongside each other, like in the other proposal I put forward. 

> Maybe. I just didn't catch a good argument in favor of the added complexity of carrying the serialized version "on the side" (more structures that will never be written outside of block reception/creation, and/or more function arguments, more "dots" resulting in stuff like stored_block.block.header.contents etc. everywhere)

I thing having a flatter architecture, with different structures for different concept living alongside each other, is a fairly general accepted principle of software engineering, so it doesn't require a "good argument" on top of it. 

See for example ["The Zen of Pyton"](https://www.python.org/dev/peps/pep-0020/), which includes "Flat is better than nested". 

Also in general, it's not the numbers of structure that increase complexity, as long as they represent different things. Actually trying to reduce the number of structure by putting things together that should be separated, is how complexity is introduced. 

So In general I think it's best to keep "the serialized form of X", separate from "X". Unless if there is a good argument to go against a general principle. 

Off course the big question is what are "different things" and what should be separated and what shouldn't be.

Yes let's discuss in more details.

**Comment by gterzian** [comment]
> That's mostly to prepare for the extra flexibility needed by the multistake protection. Some (multistaked) blocks could be considered complete even without a payload (which we could explicitly discard a posteriori or avoid downloading in some cases).

Ok, thanks for the info.

**Comment by gterzian** [comment]
> I don't like the idea of storing network messages. What if we decide to batch header transmission like we do for operations, and send different batches to different people ? What about future versioning, with different nodes with different block and network protocol versions coexisting ?

Good point.

**Comment by gterzian** [comment]
See my replies above @damip, also it's worth seeing those in the context of https://github.com/massalabs/massa/pull/2233

**Comment by damip** [comment]
> > Both work. I went for a "flat" version simply allows using a lower number of nested sub-structures.
> 
> I agree that a "flat" design would be likely superior, and the current proposal appears "nested" to me, since the serialized form appears to be nested inside the block structure. "Flat" would to me appear to require having separate concepts alongside each other, like in the other proposal I put forward.
> 
> Yes let's discuss in more details.

the same question can be raised for other things that are computed from the object contents themselves like the signature or the object's ID. I was voting for putting all the info we have on the block inside the "Block" object so we don't need to remember that some particular things like the serialized version need to be gathered in a different way than the rest of the info we have on the block.

**Comment by damip** [comment]
> > That's mostly to prepare for the extra flexibility needed by the multistake protection. Some (multistaked) blocks could be considered complete even without a payload (which we could explicitly discard a posteriori or avoid downloading in some cases).
> 
> Ok, thanks for the info. Wouldn't that be the same thing as a block with no operations? For example the information to discard to avoid further download could be stored in the header?
> 
> My point is:
> 
>     1. The current "header" appears fairly similar to a block with no payload.
> 
>     2. The current "Full block" concept appears like a block with a payload.
> 
> 
> Is that correct? And if so, what is the point of changing those concepts?

The thing with making things "evolvable" is that it's its a good practice to follow in general, and it might end up working for some things and failing for some unforeseen others, but here we are lucky to know some specifics about the future through the listing we have on the Big Refactoring, and it is important to take new design decisions with those in mind. Note that -at least here- it won't make things more complicated, but actually more evolvable.

From the point of view of network propagation and multistake protection, a block comes in two parts:
* the header + denunciations : they are propagated as an announcement
* endorsements + operations : they should be downloaded after asking explicitly, no need to send the header again as the asker has it already. Also in case of multistake they could be dropped from existing blocks and/or not downloaded.

From the point of view of consensus/graph and execution, a block has two conceptually different parts:
* header + denunciations + endorsements => required to integrate the block in the graph and compute finesses
* operations => only taken care of by the Execution module, not by consensus

We should think about both those separations and see what could be the optimal split

**Comment by gterzian** [comment]
> so we don't need to remember that some particular things like the serialized version need to be gathered in a different way than the rest of the info we have on the block.

If the serialized representation is used in different places and differently from the block itself, I think it makes sense to separate the two. My understanding is that they will indeed be used in different places, for example network will probably only need the serialized representation, which it can fetch from storage using the ids it receives in messages from other components. 

So I'm confused by the statement "need to remember that some particular things like the serialized version need to be gathered in a different way", because the point is separation of concerns. 

If indeed when using the "thing" one would always need the "other thing", and this "other thing" would have to be fetched in some complicated way that somehow would make using "the thing" more difficult, then that indeed could be a problem that one could solve by merging those concepts in one. 

In this case, I think the two concepts, the block and its serialized form, are separate and used in different ways. Hence it might make sense to keep them separate, while still keeping them "next to each other", in a flat design, as they are still related.

**Comment by gterzian** [comment]
> 

From the point of view of network propagation and multistake protection, a block comes in two parts:

    the header + denunciations : they are propagated as an announcement
    endorsements + operations : they should be downloaded after asking explicitly, no need to send the header again as the asker has it already. Also in case of multistake they could be dropped from existing blocks and/or not downloaded.

From the point of view of consensus/graph and execution, a block has two conceptually different parts:

    header + denunciations + endorsements => required to integrate the block in the graph and compute finesses
    operations => only taken care of by the Execution module, not by consensus

We should think about both those separations and see what could be the optimal split

Ok, how about we focus on moving the `Vec<Endorsement>` out of the current `BlockHeaderContent`, and replacing it with a `Vec<Denunciation>`?

**Comment by gterzian** [comment]
@damip moving the discussion about the right header/block split and potential renaming to https://github.com/massalabs/massa/issues/2264#issuecomment-1040130774

**Comment by gterzian** [comment]
> I don't like the idea of storing network messages. What if we decide to batch header transmission like we do for operations, and send different batches to different people ? What about future versioning, with different nodes with different block and network protocol versions coexisting ?

By the way I think batching sending stuff over the network is indeed a very important optimization, and that we should do this as part of the "big refactoring". 

This is something that has already been "proven" in the experience of implementing other system, see for example for Raft chapter 10 of https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf

So I agree we should not sore the network messages, but rather the serialized representations of objects, as this would allow batching multiple objects into one network message. 

Let's discuss how to do refactor the network messages in a dedicated issue.

**Comment by gterzian** [comment]
> I don't like the idea of storing network messages.

Switched to storing the serialized object, sketched over at https://github.com/massalabs/massa/pull/2305 cc @damip

**Comment by qdrn** [comment]
bump

**Comment by damip** [comment]
closing in favor of other (de-)serialization improvements, and storing the serialized version of objects
